mempool/src/tests/multi_node_test.rs

```diff
@@ -0,0 +1,941 @@
+        }
+    }
+
+    fn send_new_peer_event(
+        &mut self,
+        receiver: &PeerId,
+        new_peer: &PeerId,
+        peer_role: PeerRole,
+        origin: ConnectionOrigin,
+    ) {
+        let metadata = ConnectionMetadata::mock_with_role_and_origin(*new_peer, peer_role, origin);
+
+        let notif =
+            ConnectionNotification::NewPeer(metadata, NetworkContext::mock_with_peer_id(*receiver));
+        self.send_connection_event(receiver, notif)
+    }
+
+    fn send_lost_peer_event(&mut self, receiver: &PeerId, lost_peer: &PeerId) {
+        let notif = ConnectionNotification::LostPeer(
+            ConnectionMetadata::mock(*lost_peer),
+            NetworkContext::mock(),
+            DisconnectReason::ConnectionLost,
+        );
+        self.send_connection_event(receiver, notif)
+    }
+
+    fn send_connection_event(&mut self, peer: &PeerId, notif: ConnectionNotification) {
+        let conn_notifs_tx = self.network_conn_event_notifs_txs.get_mut(peer).unwrap();
+        conn_notifs_tx.push(*peer, notif).unwrap();
+        self.wait_for_event(peer, SharedMempoolNotification::PeerStateChange);
+    }
+
+    /// Connect two nodes, A -> B, direction is important
+    fn connect_a_to_b(&mut self, node_a: &Node, node_b: &Node) {
+        self.connect_a_to_b_with_networks(node_a, true, node_b, true);
+    }
+
+    /// Connect two nodes on specific interfaces
+    fn connect_a_to_b_with_networks(
+        &mut self,
+        node_a: &Node,
+        is_primary_a: bool,
+        node_b: &Node,
+        is_primary_b: bool,
+    ) {
+        let id_a = node_a.peer_id(is_primary_a);
+        let id_b = node_b.peer_id(is_primary_b);
+
+        let role_a = node_a.peer_role();
+        let role_b = node_b.peer_role();
+
+        // Tell B about A
+        self.send_new_peer_event(&id_b, &id_a, role_a, ConnectionOrigin::Inbound);
+
+        // Tell A about B
+        self.send_new_peer_event(&id_a, &id_b, role_b, ConnectionOrigin::Outbound);
+    }
+
+    /// Disconnect two nodes
+    fn disconnect(&mut self, node_a: &Node, node_b: &Node) {
+        let id_a = node_a.primary_peer_id();
+        let id_b = node_b.primary_peer_id();
+
+        // Tell B about A
+        self.send_lost_peer_event(&id_b, &id_a);
+
+        // Tell A about B
+        self.send_lost_peer_event(&id_a, &id_b);
+    }
+
+    /// Blocks, expecting the next event to be the type provided
+    fn wait_for_event(&mut self, peer_id: &PeerId, expected: SharedMempoolNotification) {
+        let primary_peer_id = self.primary_peer_ids.get(peer_id).unwrap_or(peer_id);
+        let subscriber = self.subscribers.get_mut(primary_peer_id).unwrap();
+
+        // TODO: This section here is the root of all flaky mempool tests, originally it would
+        // only check for 1 event, and now it checks for multiple, but the test might not end timely
+        for _ in 1..5 {
+            let event = block_on(subscriber.next()).unwrap();
```

Comment:
Yeah, we can put a timeout by doing a join with a timeout.  I'll do that because I agree it's annoying

Reply:
It's a bit more complicated since we're intertwining async and sync code here.  I tried to add a timeout, and was getting a bunch of tokio errors, so that's a different day's problem